tools/online_editor/src/index.ts

```diff
@@ -40,28 +48,209 @@ function create_demo_menu(editor: EditorWidget): Menu {
+      }
+      if (dock.parent != null) {
+        objMap.set(dock.parent, "::widget:://parent");
+      }
+      try {
+        const layout_str = JSON.stringify(layout, (_, value) => {
+          const result = objMap.get(value);
+          if (result === undefined) {
+            return value;
+          }
+          return result;
+        });
+        localStorage.setItem("layout", layout_str);
+      } catch (e) {
+        console.log("Failed to save dock layout!", e);
+      }
+    },
+  });
+
+  commands.addCommand("slint:reset-dock-layout", {
+    label: "Reset to Default Layout",
+    caption: "Reset to default layout",
+    execute: () => {
+      localStorage.removeItem("layout");
+    },
+  });
+
+  commands.addCommand("slint:restore-dock-layout", {
+    label: "Restore Dock Layout",
+    caption: "Restore the stored Dock layout",
+    execute: () => {
+      const layout_str = localStorage.getItem("layout");
+      if (layout_str != null && layout_str != "") {
+        const idMap: Map<string, Widget | Layout | null | undefined> =
+          new Map();
+        for (const id of dock_widgets.ids) {
+          idMap.set(widget_pseudo_url(id), dock_widgets.widget(id));
+        }
+        idMap.set("::widget:://DockPanel", dock);
+        idMap.set("::object:://DockLayout", dock.layout);
+        idMap.set("::object:://parent", dock.parent);
+
+        try {
+          const layout = JSON.parse(layout_str, (_, value) => {
+            const obj = idMap.get(value);
+            if (obj === undefined) {
+              // nothing we need to map!
+              return value;
+            }
+            if (obj === null) {
+              // We need to create this first!
+              return dock_widgets.create(id_from_pseudo_url(value));
+            } else {
+              return obj;
+            }
+          });
+          dock.restoreLayout(layout);
+        } catch (e) {
+          console.log("Failed to restore layout!", e);
+        }
+      }
+    },
+  });
+
+  menu.addItem({ command: "slint:save-dock-layout" });
+  menu.addItem({ command: "slint:restore-dock-layout" });
+  menu.addItem({ command: "slint:reset-dock-layout" });
+
+  for (const w of dock.widgets()) {
+    const id = w.title.label;
+    const command_name = "slint:visibility_" + id;
+    commands.addCommand(command_name, {
+      label: "Show " + w.title.label,
+      isToggled: () => {
+        return dock_widgets.widget(id) != null;
+      },
+      execute: () => {
+        const widget = dock_widgets.widget(id);
+        if (widget == null) {
+          dock_widgets.create(id);
+        } else {
+          widget.close();
+        }
+      },
+    });
+
+    menu.addItem({ command: command_name });
+  }
+
+  return menu;
+}
+
+class DockWidgets {
+  #factories: Map<string, () => Widget> = new Map();
+  #dock: DockPanel;
+
+  constructor(
+    dock: DockPanel,
+    ...factories: [() => Widget, any][] // eslint-disable-line
```

Comment:
At this point this is not a "IAddOptions" as the widgets referenced in there are replaced with their IDs. I do not have created the widgets yet, so I need that indirection.

It will fail to compile as soon as I do "as IAddOptions", so that should still be reasonably type save.

Reply:
Good Lu:)